### PR TITLE
[fx] PassManager changes

### PR DESCRIPTION
**Pull request by angelayi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81269
* #80532
* __->__ #80531

PassManager is a class used to run multiple passes on a given graph module.

Class Attributes
* `passes: List[Callable]`: A list of callable passes
* `constraints: List[Callable]`: A list of constraints
* `run_checks_after_each_pass`: Flag for running checks each pass

Class Methods:
* `__call__(graph_module: DispatchGraphModule)`:
    * Runs the passes based on the list of passes until the graph stops changes, or until `steps` number of times.
    * Each time a pass is run, it will check that the graph module still maintains the required invariants by calling `check()` and will lint the graph to check that it’s well formed if the flag `run_checks_after_each_pass` is set.
* `check(graph_module: DispatchGraphModule)`: Runs various checks on the given graph module to make sure that it contains the needed data for passes
* `add_check(check: Callable)`: Adds the `check` function to the given pass manager instance
* `add_constraint(constraint: Callable)`: Adds a constraint to the current list of constraints

We can create a PassManager and run it by doing:
```
PassManager(passes=[pass1, pass2])(graph_module)
```

Differential Revision: [D37523159](https://our.internmc.facebook.com/intern/diff/D37523159)

### CODE REVIEW

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37523159).

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37523159).

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37523159).

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37523159).

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37523159).

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37523159).

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37523159).

File: torch/fx/passes/infra/pass_manager.py [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+        time += 1
+
+    for p in passes[::-1]:
+        if not visited[p]:
+            topological_sort_util(graph, p)
+
+    res.reverse()
+    return res, circular_dependency
+
+@compatibility(is_backward_compatible=True)
```
Comment (SherlockNoMad) [inline torch/fx/passes/infra/pass_manager.py:105]: is_backward_compatible=False ?

File: torch/fx/passes/infra/pass_manager.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+    def solve_constraints(self):
+        """
+        Finds a valid traversal order based on the given constraints and orders
+        the passes based on this order.
+
+        If a circular dependency exists between the constraints and steps = 1,
+        then we will raise an error because if steps != 1 this means that we
+        will re-run the passes, allowing for circular dependencies.
+        """
+        self.passes, circular_dep = topological_sort_passes(self.passes, self.constraints)
```
Comment (SherlockNoMad) [inline torch/fx/passes/infra/pass_manager.py:212]: set self._validated = True if it we can find a valid order

File: torch/fx/passes/infra/pass_manager.py [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+
+    Args:
+        passes (Optional[List[Callable]]): List of passes. A pass is a
+            callable which modifies an object and returns modified object
+        constraint (Optional[List[Callable]]): List of constraints. A
+            constraint is a callable which takes two passes (A, B) and returns
+            True if A depends on B and False otherwise. See implementation of
+            `this_before_that_pass_constraint` for example.
+        steps (int): Max number of times we run the passes (default = 1).
+        enable_debug_pass (bool): Set to true to enable the debug passes
```
Comment (SherlockNoMad) [inline torch/fx/passes/infra/pass_manager.py:152]: not used anywhere?

File: torch/fx/passes/infra/pass_manager.py [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+            if i == j:
+                continue
+
+            for constraint in constraints:
+                if not constraint(a, b):
+                    if b in graph[a]:
+                        circular_dependency = True
+                    graph[b].append(a)
+
+    time = 1
```
Comment (SherlockNoMad) [inline torch/fx/passes/infra/pass_manager.py:78]: clean up ?

File: torch/fx/passes/infra/pass_manager.py [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+
+    # Topologically sort the graph
+    def topological_sort_util(graph, p):
```
Comment (SherlockNoMad) [inline torch/fx/passes/infra/pass_manager.py:81]: IIUC, this is a DFS based topological sort, and recursively calls itself. 
In production system, we tends avoid recursively call, as you can get into really deep calling stack if one is not careful...

I find the in_degree map traversal more understandable for topological sort, and it's probably shorter to write... 

See topo_sort in fuser_utils.py.

File: torch/fx/passes/infra/pass_manager.py [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+        # Run the set of passes `steps` number of times or until the graph stops
+        # changing
+        for _ in range(self.steps):
+            modified = False
+
+            # Run the set of passes on the graph module
+            for fn in self.passes:
+                res = fn(module)
+
+                module = res
```
Comment (SherlockNoMad) [inline torch/fx/passes/infra/pass_manager.py:258]: module = res.graph_module
modified |= res.modified

File: torch/fx/passes/infra/pass_manager.py [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+
+    def check(self, module: nn.Module) -> None:
+        pass
+
+    def __call__(self, module: nn.Module) -> nn.Module:
+        """
+        Runs a list of passes in the order based on `self.passes` on the given
+        graph module. Each time a pass is run, checks and linting will be run on
+        the graph module to ensure that it still maintains the same required
+        invariants.
```
Comment (SherlockNoMad) [inline torch/fx/passes/infra/pass_manager.py:237]: I am not sure what invariants means here?

Usually, after transformation, we want to ensure graphs are legal or valid, that there is not cycle, nodes are topologically sorted, ect... 

A more generic name for this hook could be validate() or preprocess(), or postprocess().

Comment (angelayi) [reply torch/fx/passes/infra/pass_manager.py:237]: I just meant that the function check() will get called. In the EXIR use case this function would check that specific invariants are maintained but it's not relevant here.

File: torch/fx/passes/infra/pass_manager.py (hunk 5 above)
Comment (angelayi) [reply torch/fx/passes/infra/pass_manager.py]: Updated the code to use the iterative version of topological sort! The code isn't exactly shorter due to the issue with breaking cycles, but the actual sorting part is shorter and more readable!

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37523159).

**Comment by angelayi** [comment]
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37523159).

**Review by SherlockNoMad — APPROVED** [review]
LGTM!

**Pull request closed without merging**